MRG: Adding Picard as a new ICA solver

pierreablin opened the pull request:
WIP: This PR adds the Preconditioned ICA for Real data (Picard) (see [https://arxiv.org/abs/1706.08171](https://arxiv.org/abs/1706.08171)) to the list of ICA solvers in MNE. It is faster than the current solvers, FastICA and infomax. It also converges more consistently.

This adds two dependencies to MNE:
The Picard package itself, and Numexpr which is a requirement for the package.

I have added the algorithm to the `mne/preprocessing/ica.py` file, and updated each test in `mne/preprocessing/tests/test_ica.py`.

I have also added a new test in `mne/preprocessing/tests/test_ica.py` called `test_ica_simple()` which actually tests that the ICA algorithms do their job by returning the good unmixing matrix in an experimental setup.

larsoner commented on mne/utils.py:
```diff
@@ -1070,6 +1078,7 @@ def wrapper(func):
 requires_nitime = partial(requires_module, name='nitime')
 requires_h5py = partial(requires_module, name='h5py')
 requires_numpydoc = partial(requires_module, name='numpydoc')
+requires_picard = partial(requires_module, name='picard', call=_picard_call)
```
This is never used

pierreablin replied:
Fixed :)

larsoner commented on doc/whats_new.rst:
```diff
@@ -75,6 +75,8 @@ Changelog
 
 - Add reader for manual annotations of raw data produced by Brainstorm by `Anne-Sophie Dubarry`_
 
+- Add the Picard algorithm to perform ICA, by `Pierre Ablin`_
```
```
for :class:`mne.preprocessing.ICA`
```

agramfort commented:
diff looks good to me now. @pierreablin can you add now picard in the ICA tutorial and at lease one example so people can actually find out about it?

than you'll need to update what's new. thanks

larsoner commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -58,7 +58,7 @@ def _skip_check_picard(method):
         try:
             import picard  # noqa
-        except Exception as e:
+        except:
```
always `except Exception` not bare except (unless you really mean to do so)

dengemann commented:
@pierreablin @agramfort let me know once you need a review.

jona-sassenhagen commented on mne/preprocessing/ica.py:
```diff
@@ -132,11 +132,12 @@ class ICA(ContainsMixin):
+              differences in floating point arithmetic due to exponential
+              non-linearity. Extended-Infomax seems to be more stable in this
+              respect enhancing reproducibility and stability of results.
+              Picard tends to be faster than the other algorithms.
```
Maybe we should add the references here.

pierreablin replied:
Done :)

jona-sassenhagen commented:
Very cool!

pierreablin commented:
I have changed the example `plot_run_ica.py` so that the default algorithm is picard rather than Fastica. In the tutorials about ICA, I've added a few sentences here and there to advertise for the use of this algorithm, but I haven't changed `method='fastica'`.

agramfort commented:
you have travis errors: https://travis-ci.org/mne-tools/mne-python/jobs/348338621
can you replicate @pierreablin ?

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -233,14 +234,20 @@ class ICA(ContainsMixin):
+    ----------
+    [1] Ablin, P., Cardoso, J.F., Gramfort, A., 2017. Faster Independent
+        Component Analysis by preconditioning with Hessian approximations.
+        arXiv:1706.08171
```
if we have refs for Picard we should have refs for all methods.

agramfort commented on mne/preprocessing/ica.py:
```diff
@@ -2038,12 +2055,12 @@ def _detect_artifacts(ica, raw, start_find, stop_find, ecg_ch, ecg_score_func,
-            ecg_criterion=0.1, eog_ch=None, eog_score_func='pearsonr',
-            eog_criterion=0.1, skew_criterion=-1, kurt_criterion=-1,
-            var_criterion=0, add_nodes=None, verbose=None):
+            n_pca_components=64, method='fastica', noise_cov=None,
```
this is an API break. You should put the method parameter at the end, just before verbose otherwise it's likely to break code.

agramfort commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -60,10 +60,14 @@
 # -------
 #
-# ICA parameters:
+# First, chose the ICA method. There are currently 4 possible choices:
```
chose -> choose

agramfort commented:
@jona-sassenhagen @cbrnr can you give it a try? we'd love to have your feedback.

jona-sassenhagen commented:
Hm ..?

```
ica_picard = mne.preprocessing.ICA(n_components=20, method="picard")
ica_picard.fit(raw.copy().filter(8, 35))

```


```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-81016881405e> in <module>()
      1 ica_picard = mne.preprocessing.ICA(n_components=20, method="picard")
----> 2 ica_picard.fit(raw.copy().filter(8, 35))

~/tools/mne-python/mne/preprocessing/ica.py in fit(self, inst, picks, start, stop, decim, reject, flat, tstep, reject_by_annotation, verbose)

~/tools/mne-python/mne/utils.py in verbose(function, *args, **kwargs)
    726         with use_log_level(verbose_level):
    727             return function(*args, **kwargs)
--> 728     return function(*args, **kwargs)
    729 
    730 

~/tools/mne-python/mne/preprocessing/ica.py in fit(self, inst, picks, start, stop, decim, reject, flat, tstep, reject_by_annotation, verbose)
    401             if isinstance(inst, BaseRaw):
    402                 self._fit_raw(inst, picks, start, stop, decim, reject, flat,
--> 403                               tstep, reject_by_annotation, verbose)
    404             elif isinstance(inst, BaseEpochs):
    405                 self._fit_epochs(inst, picks, decim, verbose)

~/tools/mne-python/mne/preprocessing/ica.py in _fit_raw(self, raw, picks, start, stop, decim, reject, flat, tstep, reject_by_annotation, verbose)
    469         data, self._pre_whitener = self._pre_whiten(data, raw.info, picks)
    470 
--> 471         self._fit(data, self.max_pca_components, 'raw')
    472 
    473         return self

~/tools/mne-python/mne/preprocessing/ica.py in _fit(self, data, max_pca_components, fit_type)
    616             from picard import picard
    617             _, W, _ = picard(data[:, sel].T, whiten=False,
--> 618                              random_state=random_state, **self.fit_params)
    619             del _
    620             self.unmixing_matrix_ = W

TypeError: picard() got an unexpected keyword argument 'random_state'
```

I got `picard` from here:
https://github.com/pierreablin/picard
Directly taking the git repo.

agramfort commented:
random_state was just added this week to match fastica API.

See: https://github.com/pierreablin/picard/blob/master/picard/solver.py#L17

it contains random_state.

not sure if @pierreablin pushed it to pypi yet. You may need to install
from master.

thanks

jona-sassenhagen commented:
I was confused cause I had cloned the repo ... But that was yesterday! `git pull` and it works. I will do some tests later.

pierreablin commented:
@agramfort @jona-sassenhagen I pushed picard 0.3 on pypi yesterday, where I added the random_state parameter.

I'll try to fix travis tomorrow !

jona-sassenhagen commented:
Hm. I'm not really getting speed improvements. It's about 10% slower than FastICA in my case, including on the EEGLAB sample data. Are there any particular considerations here? Preprocessing of the data? Data quantity?

pierreablin commented:
Well the most important thing is the tolerance/ max_iter. The tolerance of Picard is based on the gradient norm, while for fastica it depends on the difference between the two last iterates. When you experience slower convergence of Picard, what are the respective tol? Default?

jona-sassenhagen commented:
Yes, everything default.

jona-sassenhagen commented:
The solutions look approximately equivalently good, though I haven't done a thorough comparison.

pierreablin commented:
The thing is that the default tolerance of fastica is quite high, and roughly corresponds to the square of Picard's tolerance (i.e. if the tolerance of Picard is set to 1e-5 which is the default, you should set the tolerance of fastica to about 1e-10 to get the same decomposition), while fastica's default tolerance is 1e-4. My experience with this tolerance is that it's too high to obtain consistent results on EEG data.

jona-sassenhagen commented:
So you are saying Picard with defaults should give better/more consistent decompositions than FastICA?

Note, I am working on strongly filtered data, which may facilitate finding a good solution quickly.

pierreablin commented:
For the default tolerance of FastICA, yes. To achieve such consistence using fastica, you would have to lower the tolerance quite a bit, slowing down fastica.

cbrnr commented:
Looks good! I quickly ran it on one dataset, and indeed it is much faster than Extended Infomax, but slower than FastICA (with the default settings as noted above): Infomax: 186s, FastICA: 15s, Picard: 36s. Using the default settings, the resulting components look pretty much identical across the algorithms. The advantage of Picard isn't immediately obvious, but (1) it's always great to have lots of options, (2) I'm sure this can be found in the reference, and (3) this might also depend on the goal (ICA for ocular artifact removal vs. ICA to separate as many physiological/cortical sources as possible). I usually use Extended Infomax because this algorithm seems to perform best for the latter objective, but in reality this is based on only one paper by the authors of the algorithm. But maybe it is a good thing that results are not really that different 😄. However, the note that "Picard tends to be faster" can be confusing, since with the default settings this is clearly not the case. I'd either add more details to the note, or maybe it is time to question the defaults? If FastICA is really that inconsistent with the current default, why not adapt it?

pierreablin commented:
Even for the task of finding ECG, FastICA is not consistent (short example : https://pierreablin.github.io/picard/auto_examples/plot_ecg_detection.html#sphx-glr-auto-examples-plot-ecg-detection-py ). At least some of this lack of consistency is due to the large default tolerance. 

Now, I am clearly not an expert on the subject, but I think that the tolerance should highly depend on the type of data ICA is applied to, and on the task you are doing (as you said @cbrnr , I think it can be much higher in the case where you simply want to remove artifacts rather than obtaining a full map, since artifact sources are generally found first by the algorithms).

Now, I don't really know what to do about the default tol either for Picard or FastICA. Lowering the tolerance of FastICA would make it much slower, since it typically struggles at perfectly optimizing the underlying criterion. 

Also, I don't feel like setting the default tolerance of Picard to a high value (e.g. `tol_picard=1e-2 ` to match the tolerance of FastICA) is a good idea, since it would likely lead to the same aforementioned consistency problem.

Maybe a solution is to add a new parameter to the `ICA` class, where the user could chose a `high`, `medium`, or  `low` tolerance, depending on the task he is doing?

cbrnr commented:
@pierreablin maybe just clarify the note about Picard in this PR to get it into MNE quickly, and let's think about possible additions in a new issue? Maybe say "Picard tends to be faster than the other algorithms given suitable parameter values" or something like that.

A new parameter might be a good idea, but this depends on how it is implemented. Were you thinking of certain presets for a range of parameters (i.e. this new parameter would set `tol` and `max_iter` and possibly other related parameters to suitable values)? Alternatively, we could also describe the settings for different application scenarios more extensively in the docs (I'm usually hesitant to add yet another argument if it's not really necessary).

pierreablin commented:
> A new parameter might be a good idea, but this depends on how it is implemented. Were you thinking of certain presets for a range of parameters (i.e. this new parameter would set tol and max_iter and possibly other related parameters to suitable values)?

Exactly. Since the stopping criterion / number of iterations required to achieve tol is not the same depending on the algorithms, it would make sense to me. But it is indeed not mandatory for the moment.

For the time being, I'll just edit the doc.

larsoner commented on mne/preprocessing/ica.py:
```diff
@@ -181,8 +183,8 @@ class ICA(ContainsMixin):
-    method : {'fastica', 'infomax', 'extended-infomax'}
-        The ICA method to use. Defaults to 'fastica'.
+    method : {'fastica', 'infomax', 'extended-infomax', 'picard'}
+        The ICA method to use. Defaults to 'fastica'. For reference, see [1-4].
```
For proper sphinx/HTML links use
```
see [1]_, [2]_, [3]_, [4]_.
```
maybe there is some syntax to combine them (feel free to check) but what I wrote should work

larsoner commented on mne/preprocessing/ica.py:
```diff
@@ -233,14 +235,33 @@ class ICA(ContainsMixin):
+
+    References
+    ----------
+    [1] Hyvarinen, A., 1999. Fast and robust fixed-point algorithms for
```
See how we do it elsewhere
```
.. [1] XXX
```

larsoner commented on mne/preprocessing/ica.py:
```diff
@@ -2168,15 +2200,17 @@ def run_ica(raw, n_components, max_pca_components=100,
         If not None, override default verbose level (see :func:`mne.verbose`
         and :ref:`Logging documentation <tut_logging>` for more).
+    method : {'fastica', 'picard'}
+        The ICA method to use. Defaults to 'fastica'.
```
If we really believe your results, we should deprecate the default to `picard` (with a warning if it's not available that one should install it, and that we will fall back to `fastica`). We already do a similar warn-fallback with the mixed-norm solvers.

cbrnr replied:
Let's not get ahead of ourselves. FastICA has been around for a long time, and has been validated lots of times. Picard is very new, and I'd suggest adding it as an additional option is fine for now.

Furthermore, what happened to `'infomax'` and `'extended-infomax'` - shouldn't they be in this list?

larsoner replied:
So the belief is not there :)

cbrnr replied:
Not yet 😄 - but we shouldn't rely on belief anyway, we're not talking about religion here 😉.

larsoner commented on tutorials/plot_ica_from_raw.py:
```diff
@@ -40,7 +40,9 @@
-# Other available choices are `infomax` or `extended-infomax`
+# Other available choices are `picard`, `infomax` or `extended-infomax`.
+# Picard is generally faster and gives more consistent results than the other
+# algorithms.
```
Probably here and in the `ICA` docstring you should mention the same tradeoffs you've talked about in these PR comments, i.e. something about to get roughly equivalent convergence behavior one tolerance should be the square of another, and that if these are set equal `picard` is faster and (maybe) still more stable, etc.

cbrnr replied:
I agree, this is important. Right now it is far too general and might confuse more than explain.

larsoner commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -60,10 +60,14 @@
+# `fastica`, `picard`, `infomax` and `extended-infomax`. Picard converges
+# faster than the other algorithms, and tends to give more consistent results.
 
+method = 'fastica'
```
I'd rather change this to `picard`

cbrnr replied:
I thought this PR was about adding Picard to MNE, not making it the new default? Don't get me wrong, but IMO there should be much more testing (ideally also from many different users) happening before it gets promoted to the new default.

larsoner replied:
The change is somewhat separable from the default argument. Especially if it's faster it would be nice to have here. (Or more stable with the same speed.) As long as there is a comment as to why it's chosen for this example I think it's fine to change it here.

cbrnr replied:
I agree with @jona-sassenhagen, even if it's just for this example it's probably a bad idea to pull in extra dependencies if there is no clear benefit.

larsoner replied:
`pip`-installable (and in the MNE env) dependencies are pretty straightforward, no? One benefit (in addition to any speed or stability gains, which seem to have been argued already as potential advantages to this algorithm) is a potential increase in users who will try it and give us feedback, which sounds like a necessary feature for people to gain confidence in it..

cbrnr replied:
Right, this is not a real issue if it's just for this example. Still, can we quantify the improvement in this example? I.e., how much faster is Picard? How much more stable? Otherwise we're back to belief :smile:...

larsoner replied:
Based on previous comments, I'd expect the same speed but more reliable, or the same reliability and half the speed. We can let @pierreablin show us if he is motivated to win you over :)

larsoner replied:
Argh I mean half the time double the speed

larsoner replied:
I'm thinking a comment like this:
```
# Here we use the 'picard' solver with a permissive tolerance because ICA converges
# very quickly when finding artifacts, and 'picard' does this faster than other variants.
# For a full data decomposition, the default tolerance (1e-4) achieves roughly the same
# speed with higher reliability.
```
Would that win you over @cbrnr ?

cbrnr replied:
Yes, you can win me over with hard facts 😄. Although I doubt that this is a suitable example, because @pierreablin has mentioned that artifacts are always among the largest components, so even FastICA should consistently find them.

pierreablin replied:
I'll give it a try :)

First, look at this example (inspired by the mne tutorials): https://pierreablin.github.io/picard/auto_examples/plot_ecg_detection.html#sphx-glr-auto-examples-plot-ecg-detection-py
This is simply using ICA to find ECG artifacts. First, I use FastICA with default tol, and repeat the experiment 3 times. You can see that I get two consistent sources, for which the topography clearly matches an ECG. Then, there is a third one for two runs, which is absolutely not consistent and looks like crap, but is still marked as an ECG.

Using Picard, you always get the same ECG sources, and no additional noisy channel. On those data with the given tol/maxiter parameters, picard is also more than two times faster.


But this is only for one subject on a specific task, for a more consistent study, just look at the last figure of our paper  https://arxiv.org/pdf/1711.10873.pdf :  using the same criterion to monitor convergence, you can see that Picard ends up being much faster than Fastica, especially when it comes to solving the problem accurately (by accurately I mean with a small tolerance).

I hope that this convinces you that Picard can be useful. Now I don't advocate for using it as default anywhere, since as said before it has not been thoroughly tested by anyone other than myself.

cbrnr replied:
I never said Picard isn't useful. Actually I think it is great that it will be available in MNE! I'm just arguing that if we use it in an existing example, it should perform better in this particular example.

Now you clearly show this in your own example, although the argumentation doesn't fully convince me, because it is based on a ECG detection algorithm and not on the ICA algorithm per se. I'm not an expert in ECG detection, but the "crap" component doesn't actually look like crap - can you rule out that this is not ECG-related? What does convince me more is the fact that FastICA did not converge, therefore its components are likely to be flawed, whereas Picard apparently did converge, Picard is faster, and thus its components are better.

Getting back to the example, we're talking about an MNE example here. I think it would be important to show that Picard also performs better than FastICA in this example (showing that with the default settings FastICA doesn't converge but Picard does would be sufficient IMO). This doesn't seem to be the case, or at least there don't seem to be any problems with FastICA (although further down ECG artifacts are scored, and this yields three components, including one component you'd mark as crap).

In conclusion, if you think Picard performs better than FastICA in the MNE example, go ahead and replace FastICA! If this is not so clear like I think (but maybe I'm wrong), we should carefully consider why we'd want to replace something that does the job.

larsoner replied:
> if you think Picard performs better than FastICA in the MNE example, go ahead and replace FastICA! If this is not so clear like I think (but maybe I'm wrong), we should carefully consider why we'd want to replace something that does the job.

To me even if the two algorithms do the job equivalently in this particular case, there is value in putting `picard` here if 1) we expect it to be better in the general case and 2) we want to get people to try it. This is a valuable educational / advertising spot.

cbrnr replied:
I agree these are valid reasons. Although I'd prefer a better example that really showcases `picard` as opposed to an example where FastICA is apparently just as good. Why not add such a dedicated example? Anyway, you can go ahead and include it in the existing tutorial, I'm not strongly against it - just trying to argue from a user's perspective.

larsoner replied:
> Why not add such a dedicated example?

Hmm... true, we could create a new example and point to it from this one (in this spot about which method to use), which keeps visibility high. @pierreablin you could translate the example from your other repo over here as a tutorial to convince people. Then if we ever decide to change the default from FastICA to picard, we change the method here. @pierreablin WDYT?

jona-sassenhagen commented:
Even if Picard is clearly and universally superior, I think people still expect a true and tried algorithm.

Also: Picard required an external repo. Our infomax is native/fastica depends on sklearn.

________________________________
From: Clemens Brunner <notifications@github.com>
Sent: Monday, March 5, 2018 4:44:33 PM
To: mne-tools/mne-python
Cc: jona-sassenhagen; Mention
Subject: Re: [mne-tools/mne-python] WIP: Adding Picard as a new ICA solver (#4978)


@cbrnr commented on this pull request.

________________________________

In mne/preprocessing/ica.py<https://github.com/mne-tools/mne-python/pull/4978#discussion_r172229290>:

> @@ -2168,15 +2200,17 @@ def run_ica(raw, n_components, max_pca_components=100,
     verbose : bool, str, int, or None
         If not None, override default verbose level (see :func:`mne.verbose`
         and :ref:`Logging documentation <tut_logging>` for more).
+    method : {'fastica', 'picard'}
+        The ICA method to use. Defaults to 'fastica'.


Let's not get ahead of ourselves. FastICA has been around for a long time, and has been validated lots of times. Picard is very new, and I'd suggest adding it as an additional option is fine for now.

Furthermore, what happened to 'infomax' and 'extended-infomax' - shouldn't they be in this list?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mne-tools/mne-python/pull/4978#discussion_r172229290>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEHyIpBEyJMJ-A1dUpVKKIMNo8BlzVLrks5tbV1cgaJpZM4SXDW8>.

jona-sassenhagen commented:
That would still mean running the example depends on Picard as an additional external dependency.

________________________________
From: Eric Larson <notifications@github.com>
Sent: Monday, March 5, 2018 4:52:07 PM
To: mne-tools/mne-python
Cc: jona-sassenhagen; Mention
Subject: Re: [mne-tools/mne-python] WIP: Adding Picard as a new ICA solver (#4978)


@larsoner commented on this pull request.

________________________________

In tutorials/plot_artifacts_correction_ica.py<https://github.com/mne-tools/mne-python/pull/4978#discussion_r172232173>:

>
+method = 'fastica'


The change is somewhat separable from the default argument. Especially if it's faster it would be nice to have here. (Or more stable with the same speed.) As long as there is a comment as to why it's chosen for this example I think it's fine to change it here.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mne-tools/mne-python/pull/4978#discussion_r172232173>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEHyItV0SoRWMlXD0p5h3JTe9fPDTyASks5tbV8hgaJpZM4SXDW8>.

larsoner commented:
You need to update `circleci.yml`:
```

Unexpected failing examples:
/home/ubuntu/mne-python/examples/preprocessing/plot_run_ica.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/ubuntu/mne-python/examples/preprocessing/plot_run_ica.py", line 49, in <module>
    ica = ICA(n_components=0.95, method='picard').fit(epochs)
  File "<string>", line 2, in fit
  File "/home/ubuntu/mne-python/mne/utils.py", line 728, in verbose
    return function(*args, **kwargs)
  File "/home/ubuntu/mne-python/mne/preprocessing/ica.py", line 426, in fit
    self._fit_epochs(inst, picks, decim, verbose)
  File "/home/ubuntu/mne-python/mne/preprocessing/ica.py", line 530, in _fit_epochs
    self._fit(data, self.max_pca_components, 'epochs')
  File "/home/ubuntu/mne-python/mne/preprocessing/ica.py", line 637, in _fit
    from picard import picard
ImportError: No module named picard
```

agramfort commented:
picard is added to environment2.yml it's weird

larsoner commented:
> picard is added to environment2.yml it's weird

Can you elaborate? We try to keep the two environments similar (2 staying stock conda, 3 staying as close to stock as we can while having all deps fulfilled).

agramfort commented:
circle.yml uses environment2.yml
same as travis with conda. So if it works on travis I don't understand why
it fails on circle.

larsoner commented:
Run the CircleCI build without cache. It caches the conda env

agramfort commented:
guys I am biased here so I tend to shut up

but I enjoy a look reading this conversation. I think it does really help
@pierreablin to
have strong arguments in terms of neuroscience data analysis. thanks a lot
for this.

agramfort commented:
I did not want to mix tutorials with research results. Tutorials should be
about learning the tools. If you want to have this communication /
convincing example in mne we'll need a different example.

who thinks it should live in mne ? or not picard?

jona-sassenhagen commented:
We have a tutorial where Stockwell is compared to multitaper and wavelets. That feels a bit similar.

cbrnr commented:
I guess this is exactly the point I was trying to make - Picard is still research, whereas FastICA has been around for a long time. In a tutorial, I would use the standard algorithm. But I think having a dedicated example comparing different ICA algorithms including a nice discussion would be interesting.

agramfort commented:
ok so we should mention picard in the tutorial and point to an advanced
example we could put in examples/preprocessing
so people can click and be convinced

jona-sassenhagen commented:
Makes sense to me.

larsoner commented:
I think something along the lines of
```
.. note:: The default in MNE is ... but picard ... see :ref:`sphx_glr_auto_examples_...`
```
in the main ICA tutorial pointing to a convincing pedagogical example is a good plan.

cbrnr commented:
Sounds good!

larsoner commented:
@pierreablin it seems like this is a couple of ReST tweaks away from being done. Do you have time to work on it soon-ish to make 0.16?

pierreablin commented:
I'm working on the example, I'm struggling to find a very convincing one. The rest should be ready pretty soon!

pierreablin commented:
I think for now on I'll stick to a simple example, just so that people know picard is here.

larsoner commented:
Is this MRG then?

pierreablin commented:
Yes :)

larsoner commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,72 @@
+#
+# ###############################################################################
+# # Extended-infomax
+run_ica('extended-infomax')
```
Approximately how long does this take? If it's <1 minute it's fine, if it's 10 minutes it's not.

larsoner replied:
I guess I can tell by the timings on CircleCI's rendering that it's about 3 minutes. I guess this is okay for an ICA example on real data, which more or less has to be computationally intensive.

pierreablin replied:
It's <1 min on my machine ! very few components.

larsoner commented on mne/preprocessing/tests/test_ica.py:
```diff
@@ -796,8 +849,7 @@ def test_n_components_and_max_pca_components_none():
+    ica = ICA(max_pca_components=max_pca_components, method=method,
               n_components=n_components, random_state=random_state)
     with warnings.catch_warnings(record=True):  # convergence
         ica.fit(epochs)
```
Did these changes add much time to the testing compared to `master`? A few sec is okay (for this type of functionality), a minute is not (unless you've fully optimized it for speed already and there is no way to make it better without compromising the tests).

pierreablin replied:
By "these changes", do you mean testing also for picard? If yes, then it roughly doubles the testing time. 

I tried to speed up the tests a little bit here and there, but I don't think it's safe not to test picard in each setting?

larsoner replied:
Yes I mean all the `tests` changes -- what is the rough overall change, then (10 sec, a minute)?

pierreablin replied:
On my machine, the updated full test takes ~ 30 sec to run (so it adds ~ 15 sec).

Also, there is a slow test which takes about half of the time.

larsoner replied:
Okay thanks. That is a lot but I think we can live with it for ICA. Someday maybe we can speed it up with fewer channels, data points, etc.

larsoner commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -56,14 +56,24 @@
+# `fastica`, `picard`, `infomax` and `extended-infomax`.
+#
+# .. note:: The default method in MNE is FastICA, which along with Infomax is
+#           one of the most widely used ICA algorithm. Picard is a
```
algorithm - > algorithms

larsoner commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -56,14 +56,24 @@
+#           one of the most widely used ICA algorithm. Picard is a
+#           new algorithm that is expected to converge faster than FastICA and
+#           Infomax, especially when the aim is to recover accurate maps with
+#           a low tolerance parameter.
```
"... parameter, see [1]_ for more information." and add `References` section at the bottom with your paper. Even a preprint is okay for now, we can update it later.

larsoner commented on tutorials/plot_ica_from_raw.py:
```diff
@@ -39,8 +39,14 @@
 # 1) Fit ICA model using the FastICA algorithm.
-
-# Other available choices are `infomax` or `extended-infomax`
+# Other available choices are `picard`, `infomax` or `extended-infomax`.
```
double-backticks on these names

larsoner commented on tutorials/plot_ica_from_raw.py:
```diff
@@ -39,8 +39,14 @@
+#           one of the most widely used ICA algorithm. Picard is a
+#           new algorithm that is expected to converge faster than FastICA and
+#           Infomax, especially when the aim is to recover accurate maps with
+#           a low tolerance parameter.
```
Same comment about a ref.

larsoner commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,72 @@
+"""
+===========================================
+Compare the different ICA algorithms in MNE
+===========================================
```
Please add this path to `doc/documentation.rst` in the `preprocessing` section

larsoner commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,72 @@
+# Picard
+run_ica('picard')
+
+# ###############################################################################
```
These spaces `# ####` break the formatting:

https://7471-1301584-gh.circle-artifacts.com/0/HTML/html/auto_examples/preprocessing/plot_ica_comparison.html

cbrnr commented:
I'm not sure if this is intended, but the output of all algorithms look the same (except for the sign):

![screen shot 2018-03-29 at 14 25 26](https://user-images.githubusercontent.com/4377312/38088728-247b3e0c-335d-11e8-8e29-92c510b1c03f.png)

pierreablin commented:
It is intended, the example is just here to show that there are other algorithms than fastica in MNE.

I find it difficult to find a simple example where picard is clearly better than fastica for the task of extracting artifacts, since artifacts are "easy to find".

I have to come up with a convincing example of the superiority of picard for a practician, but for the moment I think it's ok to show that it works.

cbrnr commented:
The example is rather specific for showcasing all ICA algorithms. IMO it would be sufficient to show the first 20 components of all algorithms and completely skip the ECG artifact detection part. Then it is perfectly fine that the results are actually pretty much identical! You could note the run time for each algorithm in the console as valuable additional information.

pierreablin commented:
You're right, I have changed this!

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+"""
+
+# Authors: Pierre Ablin <pierreablin@gmail.com>
+#
```
Remove this blank line (there should be only 1).

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+#
+#
+# License: BSD (3-clause)
+from time import time
```
Insert blank line after comment and before import

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+# Read and preprocess the data. Preprocessing consists of:
+#
+# - EEG channel selection
+#
```
Remove this blank line

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+
+raw = mne.io.read_raw_fif(raw_fname, preload=True)
+
+picks = mne.pick_types(raw.info, meg=True, eeg=False, eog=False,
```
Shorten to `picks = mne.pick_types(raw.info)` (you are using the defaults anyway).

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+
+def run_ica(method):
+    ica = ICA(n_components=20, method=method,
+              random_state=0)
```
Put on previous line

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+    t0 = time()
+    ica.fit(raw, picks=picks, reject=reject)
+    fit_time = time() - t0
+    title = ('ICA decomposition using %s. Took %.1f'
```
Shorten to `title = ('ICA decomposition using %s (took %.1fs)' % (method, fit_time))`.

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+run_ica('picard')
+
+###############################################################################
+#  Infomax
```
Should be `# Infomax` (remove one space after the `#`)

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,69 @@
+run_ica('infomax')
+
+###############################################################################
+# Extended-infomax
```
Should be `# Extended Infomax` (without the `-` and capital `I`)

cbrnr commented on mne/preprocessing/ica.py:
```diff
@@ -133,11 +133,16 @@ class ICA(ContainsMixin):
-              enhancing reproducibility and stability of results.
+    .. note:: Methods implemented are FastICA (default), Infomax,
+              Extended-Infomax, and Picard. Infomax can be quite sensitive to
+              differences in floating point arithmetic due to exponential
```
Are we 100% sure it is due to the exponential non-linearity? Maybe just remove "due to exponential non-linearity".

pierreablin replied:
Yes I don't think that's the reason either, it's just been in the note since the beginning.

cbrnr replied:
I know, now is a good time to get rid of it :smile:

cbrnr commented on mne/preprocessing/ica.py:
```diff
@@ -133,11 +133,16 @@ class ICA(ContainsMixin):
+              differences in floating point arithmetic due to exponential
+              non-linearity. Extended-Infomax seems to be more stable in this
+              respect enhancing reproducibility and stability of results.
+              The stopping criterion of FastICA, Infomax, Extended
```
criteria

cbrnr commented on mne/preprocessing/ica.py:
```diff
@@ -133,11 +133,16 @@ class ICA(ContainsMixin):
+              respect enhancing reproducibility and stability of results.
+              The stopping criterion of FastICA, Infomax, Extended
+              Infomax and Picard differ, making it hard to compare different
+              tolerances level, but a rule of thumb is
```
tolerance levels

cbrnr commented on mne/preprocessing/ica.py:
```diff
@@ -133,11 +133,16 @@ class ICA(ContainsMixin):
+              Infomax and Picard differ, making it hard to compare different
+              tolerances level, but a rule of thumb is
+              `tol_fastica = tol_picard ** 2`. Reducing the tolerance speeds up
+              estimation but lowers results consistency.
```
but the consistency of the obtained results decreases

cbrnr commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -56,14 +56,24 @@
 # -------
 #
-# ICA parameters:
+# First, choose the ICA method. There are currently 4 possible choices:
```
4 -> four

cbrnr commented on tutorials/plot_artifacts_correction_ica.py:
```diff
@@ -56,14 +56,24 @@
+
+method = 'fastica'
 
+# Chose other parameters
```
Choose

cbrnr commented:
After my minor issues, +1 for merge!

cbrnr commented on examples/preprocessing/plot_ica_comparison.py:
```diff
@@ -0,0 +1,65 @@
+###############################################################################
+# Read and preprocess the data. Preprocessing consists of:
+# - EEG channel selection
+#
```
Can you remove this blank line here or is it necessary?

larsoner commented on examples/preprocessing/plot_run_ica.py:
```diff
@@ -23,11 +23,8 @@
 
 ###############################################################################
 # Read and preprocess the data. Preprocessing consists of:
-#
```
Turns out that this line was necessary

https://7490-1301584-gh.circle-artifacts.com/0/html/auto_examples/preprocessing/plot_run_ica.html

larsoner replied:
(the ones in between the `-` lines shouldn't be, though)

FYI I also did a merge commit to resolve a conflict before noticing this

cbrnr commented on examples/preprocessing/plot_run_ica.py:
```diff
@@ -23,8 +23,11 @@
 # Read and preprocess the data. Preprocessing consists of:
+#
 # - MEG channel selection
+#
```
As @larsoner pointed out, I think blank lines are only required before and after a list - not in between the list items...

larsoner commented:
Beautiful commit messages @pierreablin :)

larsoner commented:
I'll merge this once the CIs come back happy and I see that the `-` lines are properly converted to a bulleted list.

larsoner commented:
All better

https://7497-1301584-gh.circle-artifacts.com/0/html/auto_examples/preprocessing/plot_run_ica.html

larsoner commented:
So you are happy with the updated example, then? 

https://7499-1301584-gh.circle-artifacts.com/0/html/auto_examples/preprocessing/plot_ica_comparison.html

jona-sassenhagen commented:
I think it looks good.

cbrnr commented:
I like the new example! There's an `ImportError` in one of the Python 2.7 environments, did you add `picard` to the list of requirements?

cbrnr commented on .travis.yml:
```diff
@@ -49,7 +49,7 @@ matrix:
         - os: linux
           env: PYTHON_VERSION=2.7
-               CONDA_DEPENDENCIES="numpy=1.8 scipy=0.12 matplotlib=1.3 pandas=0.13 scikit-learn=0.15 nose pytest pytest-cov"
+               CONDA_DEPENDENCIES="numpy=1.8 scipy=0.12 matplotlib=1.3 pandas=0.13 scikit-learn=0.15 nose pytest pytest-cov python-picard numexpr"
```
I think you need to put `python-picard` into `PIP_DEPENDENCIES` because it's not available in conda.

pierreablin replied:
yes indeed! thanks

larsoner commented on .travis.yml:
```diff
@@ -49,7 +49,8 @@ matrix:
           env: PYTHON_VERSION=2.7
-               CONDA_DEPENDENCIES="numpy=1.8 scipy=0.12 matplotlib=1.3 pandas=0.13 scikit-learn=0.15 nose pytest pytest-cov"
+               CONDA_DEPENDENCIES="numpy=1.8 scipy=0.12 matplotlib=1.3 pandas=0.13 scikit-learn=0.15 nose pytest pytest-cov numexpr"
+               PIP_DEPENDENCIES="python-picard"
```
This should not be necessary. The failure on the Travis run from my commit is indicative that if someone has `sklearn` installed but not `picard` then tests are not properly skipped. It would be better to leave this run alone (so it is in a has-sklearn does-not-have-picard state) and fix that problem.

You can debug locally easily by temporarily removing `picard` from your system.

pierreablin replied:
Thanks for the tip! I think it's fixed now :)

larsoner commented:
Thanks @pierreablin. Now everybody go try it!
